### PR TITLE
Fix year 0 is out of range in test_from_json_struct_timestamp 

### LINKED ISSUE #9936: [BUG] Nightly CI of non-UTC time zone reports 'year 0 is out of range' error
**Describe the bug**
In non-UTC time zone CI, it reports:
```
[2023-12-04T03:24:48.506Z] =========================== short test summary info ============================

[2023-12-04T03:24:48.506Z] FAILED ../../src/main/python/json_test.py::test_from_json_struct_timestamp[True-LEGACY--"[ \t\xa0\u1680\u180e\u2000-\u200a\u202f\u205f\u3000]?[0-9]{4}[ \t\xa0\u1680\u180e\u2000-\u200a\u202f\u205f\u3000]?"][DATAGEN_SEED=1701654666, ALLOW_NON_GPU(ProjectExec)] - ValueError: year 0 is out of range

[2023-12-04T03:24:48.506Z] FAILED ../../src/main/python/json_test.py::test_from_json_struct_timestamp[True-LEGACY-yyyy-MM-dd'T'HH:mm:ss[.SSS][XXX]-"[ \t\xa0\u1680\u180e\u2000-\u200a\u202f\u205f\u3000]?[0-9]{4}[ \t\xa0\u1680\u180e\u2000-\u200a\u202f\u205f\u3000]?"][DATAGEN_SEED=1701654666, ALLOW_NON_GPU(ProjectExec)] - ValueError: year 0 is out of range

[2023-12-04T03:24:48.506Z] FAILED ../../src/main/python/json_test.py::test_from_json_struct_timestamp[True-LEGACY-yyyy-MM-dd'T'HH:mm:ss.SSSXXX-"[ \t\xa0\u1680\u180e\u2000-\u200a\u202f\u205f\u3000]?[0-9]{4}[ \t\xa0\u1680\u180e\u2000-\u200a\u202f\u205f\u3000]?"][DATAGEN_SEED=1701654666, INJECT_OOM, ALLOW_NON_GPU(ProjectExec)] - ValueError: year 0 is out of range

[2023-12-04T03:24:48.506Z] FAILED ../../src/main/python/json_test.py::test_from_json_struct_timestamp[True-LEGACY-dd/MM/yyyy'T'HH:mm:ss[.SSS][XXX]-"[ \t\xa0\u1680\u180e\u2000-\u200a\u202f\u205f\u3000]?[0-9]{4}[ \t\xa0\u1680\u180e\u2000-\u200a\u202f\u205f\u3000]?"][DATAGEN_SEED=1701654666, ALLOW_NON_GPU(ProjectExec)] - ValueError: year 0 is out of range

[2023-12-04T03:24:48.506Z] FAILED ../../src/main/python/json_test.py::test_from_json_struct_timestamp[True-CORRECTED--"[ \t\xa0\u1680\u180e\u2000-\u200a\u202f\u205f\u3000]?[0-9]{4}[ \t\xa0\u1680\u180e\u2000-\u200a\u202f\u205f\u3000]?"][DATAGEN_SEED=1701654666, INJECT_OOM, ALLOW_NON_GPU(ProjectExec)] - ValueError: year 0 is out of range

[2023-12-04T03:24:48.506Z] FAILED ../../src/main/python/json_test.py::test_from_json_struct_timestamp[True-CORRECTED-yyyy-MM-dd'T'HH:mm:ss[.SSS][XXX]-"[ \t\xa0\u1680\u180e\u2000-\u200a\u202f\u205f\u3000]?[0-9]{4}[ \t\xa0\u1680\u180e\u2000-\u200a\u202f\u205f\u3000]?"][DATAGEN_SEED=1701654666, INJECT_OOM, ALLOW_NON_GPU(ProjectExec)] - ValueError: year 0 is out of range

[2023-12-04T03:24:48.507Z] FAILED ../../src/main/python/json_test.py::test_from_json_struct_timestamp[True-CORRECTED-yyyy-MM-dd'T'HH:mm:ss.SSSXXX-"[ \t\xa0\u1680\u180e\u2000-\u200a\u202f\u205f\u3000]?[0-9]{4}[ \t\xa0\u1680\u180e\u2000-\u200a\u202f\u205f\u3000]?"][DATAGEN_SEED=1701654666, INJECT_OOM, ALLOW_NON_GPU(ProjectExec)] - ValueError: year 0 is out of range

[2023-12-04T03:24:48.507Z] FAILED ../../src/main/python/json_test.py::test_from_json_struct_timestamp[True-CORRECTED-dd/MM/yyyy'T'HH:mm:ss[.SSS][XXX]-"[ \t\xa0\u1680\u180e\u2000-\u200a\u202f\u205f\u3000]?[0-9]{4}[ \t\xa0\u1680\u180e\u2000-\u200a\u202f\u205f\u3000]?"][DATAGEN_SEED=1701654666, ALLOW_NON_GPU(ProjectExec)] - ValueError: year 0 is out of range

[2023-12-04T03:24:48.507Z] FAILED ../../src/main/python/json_test.py::test_from_json_struct_timestamp[False-LEGACY--"[ \t\xa0\u1680\u180e\u2000-\u200a\u202f\u205f\u3000]?[0-9]{4}[ \t\xa0\u1680\u180e\u2000-\u200a\u202f\u205f\u3000]?"][DATAGEN_SEED=1701654666, ALLOW_NON_GPU(ProjectExec)] - ValueError: year 0 is out of range

[2023-12-04T03:24:48.507Z] FAILED ../../src/main/python/json_test.py::test_from_json_struct_timestamp[False-LEGACY-yyyy-MM-dd'T'HH:mm:ss[.SSS][XXX]-"[ \t\xa0\u1680\u180e\u2000-\u200a\u202f\u205f\u3000]?[0-9]{4}[ \t\xa0\u1680\u180e\u2000-\u200a\u202f\u205f\u3000]?"][DATAGEN_SEED=1701654666, ALLOW_NON_GPU(ProjectExec)] - ValueError: year 0 is out of range

[2023-12-04T03:24:48.507Z] FAILED ../../src/main/python/json_test.py::test_from_json_struct_timestamp[False-LEGACY-yyyy-MM-dd'T'HH:mm:ss.SSSXXX-"[ \t\xa0\u1680\u180e\u2000-\u200a\u202f\u205f\u3000]?[0-9]{4}[ \t\xa0\u1680\u180e\u2000-\u200a\u202f\u205f\u3000]?"][DATAGEN_SEED=1701654666, ALLOW_NON_GPU(ProjectExec)] - ValueError: year 0 is out of range

[2023-12-04T03:24:48.507Z] FAILED ../../src/main/python/json_test.py::test_from_json_struct_timestamp[False-LEGACY-dd/MM/yyyy'T'HH:mm:ss[.SSS][XXX]-"[ \t\xa0\u1680\u180e\u2000-\u200a\u202f\u205f\u3000]?[0-9]{4}[ \t\xa0\u1680\u180e\u2000-\u200a\u202f\u205f\u3000]?"][DATAGEN_SEED=1701654666, INJECT_OOM, ALLOW_NON_GPU(ProjectExec)] - ValueError: year 0 is out of range

[2023-12-04T03:24:48.507Z] FAILED ../../src/main/python/json_test.py::test_from_json_struct_timestamp[False-CORRECTED--"[ \t\xa0\u1680\u180e\u2000-\u200a\u202f\u205f\u3000]?[0-9]{4}[ \t\xa0\u1680\u180e\u2000-\u200a\u202f\u205f\u3000]?"][DATAGEN_SEED=1701654666, ALLOW_NON_GPU(ProjectExec)] - ValueError: year 0 is out of range

[2023-12-04T03:24:48.507Z] FAILED ../../src/main/python/json_test.py::test_from_json_struct_timestamp[False-CORRECTED-yyyy-MM-dd'T'HH:mm:ss[.SSS][XXX]-"[ \t\xa0\u1680\u180e\u2000-\u200a\u202f\u205f\u3000]?[0-9]{4}[ \t\xa0\u1680\u180e\u2000-\u200a\u202f\u205f\u3000]?"][DATAGEN_SEED=1701654666, ALLOW_NON_GPU(ProjectExec)] - ValueError: year 0 is out of range

[2023-12-04T03:24:48.507Z] FAILED ../../src/main/python/json_test.py::test_from_json_struct_timestamp[False-CORRECTED-yyyy-MM-dd'T'HH:mm:ss.SSSXXX-"[ \t\xa0\u1680\u180e\u2000-\u200a\u202f\u205f\u3000]?[0-9]{4}[ \t\xa0\u1680\u180e\u2000-\u200a\u202f\u205f\u3000]?"][DATAGEN_SEED=1701654666, INJECT_OOM, ALLOW_NON_GPU(ProjectExec)] - ValueError: year 0 is out of range

[2023-12-04T03:24:48.507Z] FAILED ../../src/main/python/json_test.py::test_from_json_struct_timestamp[False-CORRECTED-dd/MM/yyyy'T'HH:mm:ss[.SSS][XXX]-"[ \t\xa0\u1680\u180e\u2000-\u200a\u202f\u205f\u3000]?[0-9]{4}[ \t\xa0\u1680\u180e\u2000-\u200a\u202f\u205f\u3000]?"][DATAGEN_SEED=1701654666, INJECT_OOM, ALLOW_NON_GPU(ProjectExec)] - ValueError: year 0 is out of range

[2023-12-04T03:24:48.507Z] = 16 failed, 19637 passed, 2021 skipped, 371 xfailed, 403 xpassed, 915 warnings in 5621.07s (1:33:41) =
```
It's because string regexp produce not-supported timestamp string.

**Expected behavior**
It's blocking branch-23.12, so fix seed can fix fast, then handle it later properly.

**Environment details (please complete the following information)**

**Additional context**

### PR DESCRIPTION
**Pull request by thirtiseven**
Fixes #9936 

Add regex to generate 0002 to 9999:
```
([0-9]{3}[2-9]|([1-9][0-9]{2}|0[1-9][0-9]|00[1-9])[0-1])
```

Breakdown:
```
[0-9]{4} = [0-9]{3}[2-9] + [0-9]{3}[0-1]

[0-9]{3}[0-1] = [1-9][0-9]{2}[0-1] + 0[0-9]{2}[0-1]

0[0-9]{2}[0-1] = 0[1-9][0-9][0-1] + 00[0-9][0-1]

00[0-9][0-1] = 00[1-9][0-1] + 000[0-1] (we don't want this)

so the left part is:

 [0-9]{3}[2-9] + [1-9][0-9]{2}[0-1] + 0[1-9][0-9][0-1] + 00[1-9][0-1]
= ([0-9]{3}[2-9]|([1-9][0-9]{2}|0[1-9][0-9]|00[1-9])[0-1])
```

Verified it's correct by checking all values with `sre_yield`.

This PR also moves similar regexps to data_gen.py for reuse.

### CODE REVIEW

**Comment by thirtiseven** [comment]
About the 'year 0 is out of range' error, I checked all StringGen usages in our IT, here are some suspicious cases:
- `test_cast_string_date_fallback`
- all cases use `str_date_and_format_gen`
- `test_explain_set_config`

I will try to test and fix them, but maybe not here because they can't be solved by the regexp we already have.

File: integration_tests/src/main/python/data_gen.py [hunk 1]
```diff
@@ -1200,4 +1200,10 @@ def get_25_partitions_df(spark):
 # This will be deprecated and replaced case specified non GPU allow list
 non_utc_allow = ['ProjectExec', 'FilterExec', 'FileSourceScanExec', 'BatchScanExec', 'CollectLimitExec',
                  'DeserializeToObjectExec', 'DataWritingCommandExec', 'WriteFilesExec', 'ShuffleExchangeExec',
-                 'ExecutedCommandExec'] if is_not_utc() else []
\ No newline at end of file
+                 'ExecutedCommandExec'] if is_not_utc() else []
+
+# date related regexps for generating date strings within python's range limits
+
+date_start_1_2_1 = '(0{0,3}1-(0?[2-9]|[1-3][0-9]))|(([0-9]{0,3}[2-9]|[1-9][0-9]{0,2}[01])-[0-3]?[0-9])-[0-5]?[0-9]'
```
Comment (res-life) [inline integration_tests/src/main/python/data_gen.py:1207]: Add comment like:  regexp to generate date from 0001-02-01

File: integration_tests/src/main/python/data_gen.py [hunk 2]
```diff
@@ -1200,4 +1200,10 @@ def get_25_partitions_df(spark):
...
+
+yyyy_start_0003 = '([0-9]{3}[2-9]|([1-9][0-9]{2}|0[1-9][0-9]|00[1-9])[0-1])'
```
Comment (res-life) [inline integration_tests/src/main/python/data_gen.py:1209]: start from 0002 or 0003?
Add comment like: regexp to generate year from 0002, forat is yyyy

File: integration_tests/src/main/python/data_gen.py (hunk 1 above)
Comment (thirtiseven) [reply integration_tests/src/main/python/data_gen.py:1208]: done.

File: integration_tests/src/main/python/data_gen.py (hunk 2 above)
Comment (thirtiseven) [reply integration_tests/src/main/python/data_gen.py]: nice catch, done.

**Comment by res-life** [comment]
LGTM

**Review by res-life — APPROVED** [review]

**Comment by thirtiseven** [comment]
build

File: integration_tests/src/main/python/data_gen.py [hunk 3]
```diff
@@ -1200,4 +1200,12 @@ def get_25_partitions_df(spark):
...
-                 'ExecutedCommandExec'] if is_not_utc() else []
\ No newline at end of file
+                 'ExecutedCommandExec'] if is_not_utc() else []
+
+# date related regexps for generating date strings within python's range limits
+
+# regexp to generate date from 0001-02-01, format is yyyy-MM-dd
+date_start_1_2_1 = '(0{0,3}1-(0?[2-9]|[1-3][0-9]))|(([0-9]{0,3}[2-9]|[1-9][0-9]{0,2}[01])-[0-3]?[0-9])-[0-5]?[0-9]'
+
+# regexp to generate year from 0002, forat is yyyy
```
Comment (winningsix) [inline integration_tests/src/main/python/data_gen.py:1210]: typo: ```format```

Comment (winningsix) [reply integration_tests/src/main/python/data_gen.py:1210]: BTW, why starting from ```0002```?

File: integration_tests/src/main/python/data_gen.py [hunk 4]
```diff
@@ -1200,4 +1200,12 @@ def get_25_partitions_df(spark):
...
+yyyy_start_0002 = '([0-9]{3}[2-9]|([1-9][0-9]{2}|0[1-9][0-9]|00[1-9])[0-1])'
```
Comment (sperlingxx) [inline integration_tests/src/main/python/data_gen.py:1211]: nit: empty tailing line is missed

**Comment by thirtiseven** [comment]
premerge failed because of the rounding necessary issue.

File: integration_tests/src/main/python/data_gen.py (hunk 3 above)
Comment (thirtiseven) [reply integration_tests/src/main/python/data_gen.py:1210]: I think 0001 will be convert to 0001-01-01 as timestamp in this case, which is out of range.

File: integration_tests/src/main/python/data_gen.py (hunk 4 above)
Comment (thirtiseven) [reply integration_tests/src/main/python/data_gen.py]: done

**Comment by thirtiseven** [comment]
build

**Comment by thirtiseven** [comment]
rounding necessary issue again...

**Comment by thirtiseven** [comment]
build

**Review by winningsix — APPROVED** [review]

**Pull request merged**